dsl_parser/elements/data_types.py

```diff
@@ -0,0 +1,203 @@
+    Element,
+    Dict,
+    DictElement,
+    Leaf,
+    StringElement)
+from dsl_parser.framework.requirements import (
+    Value,
+    Requirement,
+    sibling_predicate)
+
+
+class SchemaPropertyDescription(Element):
+
+    schema = Leaf(type=str)
+
+
+class SchemaPropertyType(Element):
+
+    schema = Leaf(type=str)
+
+    # requires will be modified later.
+    requires = {}
+
+    provides = ['component_types']
+
+    def validate(self, data_type, component_types):
+        if self.initial_value and self.initial_value not in \
+                constants.USER_PRIMITIVE_TYPES and not data_type:
+            raise exceptions.DSLParsingLogicException(
+                exceptions.ERROR_UNKNOWN_TYPE,
+                "Illegal type name '{0}'".format(self.initial_value))
+
+    def calculate_provided(self, data_type, component_types):
+        component_types = component_types or {}
+        if self.value and self.value not in constants.USER_PRIMITIVE_TYPES:
+            component_types = copy.copy(component_types)
+            component_types[self.value] = data_type
+        return {'component_types': component_types}
+
+
+class SchemaPropertyDefault(Element):
+
+    schema = Leaf(type=elements.PRIMITIVE_TYPES)
+
+    requires = {
+        SchemaPropertyType: [
+            Value('type_name', required=False, predicate=sibling_predicate),
+            Requirement('component_types',
+                        required=False,
+                        predicate=sibling_predicate)
+        ]
+    }
+
+    def parse(self, type_name, component_types):
+        if self.initial_value is None:
+            return
+        component_types = component_types or {}
+        prop_name = self.ancestor(SchemaProperty).name
+        undefined_property_error = 'Undefined property {1} in default' \
+                                   ' value of type {0}'
+        missing_property_error = 'Property {1} is missing in default' \
+                                 ' value of type {0}'
+        current_type = self.ancestor(Schema).parent().name
+        return utils.parse_value(
+            self.initial_value,
+            type_name,
+            component_types,
+            undefined_property_error_message=undefined_property_error,
+            missing_property_error_message=missing_property_error,
+            node_name=current_type,
+            path=[prop_name]
+        )
+
+
+class SchemaProperty(Element):
+
+    schema = {
+        'default': SchemaPropertyDefault,
+        'description': SchemaPropertyDescription,
+        'type': SchemaPropertyType,
+    }
+
+    def parse(self):
+        result = self.build_dict_result()
+        return dict((k, v) for k, v in result.iteritems() if v is not None)
+
+
+class Schema(DictElement):
+
+    schema = Dict(type=SchemaProperty)
+
+
+class DataType(types.Type):
+
+    schema = {
+        constants.PROPERTIES: Schema,
+        'description': StringElement,
+        constants.DERIVED_FROM: types.DerivedFrom,
```

Comment:
I think this should be some class the derives from `types.DerivedFrom` which overrides the 
`descriptor`
https://github.com/cloudify-cosmo/cloudify-dsl-parser/blob/master/dsl_parser/elements/types.py#L58
attribute